stew/result.nim

```diff
@@ -220,45 +244,62 @@ func mapCast*[T0, E0](
   if self.isOk: result.ok(cast[T1](self.v))
   else: result.err(self.e)
 
-template `and`*(self: Result, other: untyped): untyped =
+template `and`*[T, E](self, other: Result[T, E]): Result[T, E] =
   ## Evaluate `other` iff self.isOk, else return error
   ## fail-fast - will not evaluate other if a is an error
+  ##
+  ## TODO: This API is unsafe due to potential multiple
+  ## evaluation of the `self` parameter.
   if self.isOk:
     other
   else:
     type R = type(other)
     R.err(self.e)
 
-template `or`*(self: Result, other: untyped): untyped =
+template `or`*[T, E](self, other: Result[T, E]): Result[T, E] =
   ## Evaluate `other` iff not self.isOk, else return self
   ## fail-fast - will not evaluate other if a is a value
+  ##
+  ## TODO: This API is unsafe due to potential multiple
+  ## evaluation of the `self` parameter.
   if self.isOk: self
   else: other
 
-template catch*(body: typed): Result[type(body), ref Exception] =
-  ## Convert a try expression into a Result
-  type R = Result[type(body), ref Exception]
+template catch*(body: typed): Result[type(body), ref CatchableError] =
+  ## Catch exceptions for body and store them in the Result
+  ##
+  ## ```
+  ## let r = catch: someFuncThatMayRaise()
+  ## ```
+  type R = Result[type(body), ref CatchableError]
 
   try:
     R.ok(body)
-  except:
-    R.err(getCurrentException())
-
-template capture*(T: type, e: ref Exception): Result[T, ref Exception] =
-  type R = Result[T, ref Exception]
+  except CatchableError as e:
+    R.err(e)
+
+template capture*[E: Exception](T: type, someExceptionExpr: ref E): Result[T, ref E] =
+  ## Evaluate someExceptionExpr and put the exception into a result, making sure
+  ## to capture a call stack at the capture site:
+  ##
+  ## ```
+  ## let e: Result[void, ValueError] = void.capture((ref ValueError)(msg: "test"))
+  ## echo e.error().getStackTrace()
+  ## ```
+  type R = Result[T, ref E]
 
   var ret: R
   try:
     # TODO is this needed? I think so, in order to grab a call stack, but
     #      haven't actually tested...
     if true:
       # I'm sure there's a nicer way - this just works :)
-      raise e
-  except:
-    ret = R.err(getCurrentException())
+      raise someExceptionExpr
+  except E as caught:
+    ret = R.err(caught)
   ret
 
-func `==`*(lhs, rhs: Result): bool {.inline.} =
+func `==`*[T0, E0, T1, E1](lhs: Result[T0, E0], rhs: Result[T1, E1]): bool {.inline.} =
   if lhs.isOk != rhs.isOk:
```

Comment:
Interesting level of flexibility here, I think you need to mixin `==` to make it work correctly.

Reply:
aren't mixins for templates?

the flexibility is for the same reason you want heterogeneous lookup in maps etc - ie `Result` should not have an opinion here

Reply:
`mixin` is for both generics and templates, but it has some tricky rules. If it sees an overloaded function in the current scope, it considers the symbol name as mixed in by default. This rule should probably kick in for the `==` operator actually, but it may be a good habit to always state your intentions explicitly.